Lifetime calculation in days not months

jniles opened the pull request:
This pull request attempts to clean up our inventory/lot flag calculations so that we have a uniform way of presenting the stock out, stock expiration, at risk of expiration, and at risk of stock out states across the client.

Here is what it looks like:
![image](https://user-images.githubusercontent.com/896472/104612221-40efdd00-5686-11eb-89e2-e22a9854ec44.png)

We've standardized around four "flags":
  1. exhausted - the quantity is 0.
  2. expired - the stock lot is expired
  3. near_expiration - the stock will expire before it will be consumed.
  4. at_risk - the stock is at risk of running out soon.

I've updated many of the calculations so they take place on the server will commented documentation. 

### Consumption Schedule
This PR implements what I'll call the "consumption schedule" for lots.  Basically, the lifetimes of each lot are calculated as if the enterprise was optimally using the lots within their depot.  So, the depot will use the lot that will expire soonest, then the next soonest, etc.  This means that a lot may be in stock, but will not be used until much later.

This may produce some misleading figures in the registry, which is why it may be important to show some of these calculations to the client.  I'll propose an issue about that.

### Bug Fixes
This PR also fixes bugs in the coloration of the registry and adds a final aggregation to the bottom of the registry for all four flags.

jniles committed:
fix(stock): coloration of stock lots registry

Fixes the color scheme of the stock lots registry.  Now the expired lots
show up as red always, instead of as "at risk of expiration".

Closes #5185.

jniles committed:
feat(stock): better key on stock lots registry

This commit adds all four stock states to the key at the bottom of the
registry of the lots in stock.  It also adds totals to the registry.

jniles committed:
chore: move lot flag calculations to server

Moves the lot flag calculations to the server to make it cleaner on the
client.

jniles committed:
chore(stock): clean up risk calculations

Updates the risk calculation to properly compute the risk by quantity
and time period.  Also cleans up the client so that we render in red the
quantities at risk.

jniles committed:
test(stock): update tests for consumption schedule

Updates the tests for stock now that they take into account the
consumption schedule.

jmcameron commented on server/controllers/stock/core.js:
```diff
@@ -266,23 +266,22 @@ async function getLotsDepot(depotUuid, params, finalClause) {
 
   if (_status) {
-    inventoriesWithLotsProcessed = inventoriesWithLotsProcessed.filter(row => row.status === _status);
+    inventoriesWithLotsProcessed = inventoriesWithLotsProcessed.filter(lot => lot.status === _status);
```
I like these variable name changes.  Makes it more readable!

jmcameron commented on server/controllers/stock/core.js:
```diff
@@ -692,16 +692,35 @@ async function getInventoryQuantityAndConsumption(params) {
 
-    _.map(inventoryLots, (lots) => {
+    _.forEach(inventoryLots, (lots) => {
+
```
Typo on line 713/732: hight -> higher

jmcameron commented on server/controllers/stock/core.js:
```diff
@@ -714,32 +733,80 @@ function computeLotIndicators(inventories) {
       orderedInventoryLots = _.orderBy(orderedInventoryLots, 'lifetime', 'asc');
 
       // compute the lot coefficient
```
coefficientS

jmcameron commented on server/controllers/stock/core.js:
```diff
@@ -714,32 +733,80 @@ function computeLotIndicators(inventories) {
+          // add to the running LotLifetimes so that the next product will be used after it.
+          runningLotLifetimes += lot.lifetime_lot;
+
+          // the usuable quantity remaining is the minimum of the stock actually available
```
Typo: usable

jmcameron reviewed (approved):
A minor suggestion.  We have several all-caps variables (eg, IS_IN_RISK_EXPIRATION, S_RISK, S_LOT_LIFETIME, etc).  In general, we usually reserve all-caps-with-underscores for constants or i18n language tags, not variables/flags.  I suggest making these lower-case (with underscores or camel case, as appropriate).  And maybe choosing more descriptive names as part of the change?

Question: Since the CMM is calculated by consumption over N months, it will not change very fast when stock consumption increases due to some short-term period with high usage.  However, the usage timeline depends on CMM and will not properly reflect how the usage pattern may need to change if the anomalous usage pattern continues.   We may want to consider how to address that.  I can imagine this scenario:   There is a flare up of Cholera locally.  The hospital dispenses meds for that at a higher than usual rate.   The admins start to wonder, when will we run out?  If they use the current code (as nice as it is), they may get a rosy/misleading picture.   It may be useful to provide a way to do these calculations with a more instantaneous estimate of the CMM to get a more accurate picture.  This could be done by providing a way to let the CMM period be temporarily overridden in the query.

jniles commented on server/controllers/stock/core.js:
```diff
@@ -692,16 +692,35 @@ async function getInventoryQuantityAndConsumption(params) {
+ * Computes the indicators on stock lots.  Sets the following flags:
+ *   1) expired - if the expiration date is in the past
+ *   2) exhausted - if the quantity is 0.
+ *   3) near_expiration - if the expiration is before
```
Whoops this also is an incomplete thought.

jniles committed:
chore(stock): typo fixes in comments

Cleans up the comments based on @jmcameron's feedback.

mbayopanda commented on client/src/js/services/LotService.js:
```diff
@@ -28,23 +28,24 @@ function LotService(Api, $http, util) {
+    lot.exhausted = false;
+    lot.expired = false;
+    lot.near_expiration = false;
+    lot.at_risk = false;
```
👍🏽

mbayopanda reviewed (commented):
@jniles having a lifespan of lots in days is giving our calculation accuracy, most of our flags (expired, risk of expiration, risk of stock out, stock out, etc.) coloration have an obvious understanding now than before.

My concern is about the `lot lifespan by lot`, by reading the registry we cannot understand (in an obvious way) how this column is calculated; there is no reason to have this column in the registry if the user can't understand how it is calculated.
![lot lifespan](https://user-images.githubusercontent.com/5445251/105469305-52de0b00-5c98-11eb-853f-34701d67fa15.png)

In the previous implementation, this column had a reason because it was helping us to understand some calculations just by reading the registry.

![image](https://user-images.githubusercontent.com/5445251/105470110-6fc70e00-5c99-11eb-9f90-7ec03e5f1860.png)

If our clients need that column we must have that in a way where they understand easily where it comes from and the usage just by reading the registry

If our clients cannot understand that, we must not have this column in the registry.

Or, we can add in the registry other columns for CMM in days and MS in days which can help to understand some calculations made per day; with this option, our registry will become bigger (horizontally) and less comfortable for users. With this option also if the user can't understand the columns displayed there is no reason to have these columns

dependabot-preview[bot] committed:
chore(deps-dev): bump karma from 6.0.3 to 6.0.4

Bumps [karma](https://github.com/karma-runner/karma) from 6.0.3 to 6.0.4.
- [Release notes](https://github.com/karma-runner/karma/releases)
- [Changelog](https://github.com/karma-runner/karma/blob/master/CHANGELOG.md)
- [Commits](https://github.com/karma-runner/karma/compare/v6.0.3...v6.0.4)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps): bump angular-ui-grid from 4.9.1 to 4.10.0

Bumps [angular-ui-grid](https://github.com/angular-ui/ui-grid) from 4.9.1 to 4.10.0.
- [Release notes](https://github.com/angular-ui/ui-grid/releases)
- [Changelog](https://github.com/angular-ui/ui-grid/blob/master/CHANGELOG.md)
- [Commits](https://github.com/angular-ui/ui-grid/compare/v4.9.1...v4.10.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps): bump delay from 4.4.1 to 5.0.0

Bumps [delay](https://github.com/sindresorhus/delay) from 4.4.1 to 5.0.0.
- [Release notes](https://github.com/sindresorhus/delay/releases)
- [Commits](https://github.com/sindresorhus/delay/compare/v4.4.1...v5.0.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
chore(lots): remove registry columns

Removes many of the registry columns from the stock lot registry so that
they are not as distracting to the user.

Closes #5311.

jmcameron reviewed (approved):
It looks good to me.

I really like the simplified "Lots in Stock" page.  It makes much more sense to me!

I decided to let you release this yourself since I was not sure if you wanted to rebase first.
